src/scripts/render.js

```diff
@@ -252,22 +664,293 @@ document.addEventListener("DOMContentLoaded", () => {
+    }
+  }
+
+  /**
+   * Load all the tags into the filter, called by showFilterDropdown() when user
+   * clicks on dropdown button. Gets all the tags and creates a list item for
+   * each one.
+   */
+  function loadFilterTags() {
+    const tags = notesAPI.readTags();
+    filterDropdownList.innerHTML = ""; // Clear existing items
+
+    tags.forEach((tag) => {
+      const tagItem = document.createElement("li");
+      tagItem.textContent = tag.content;
+      tagItem.style.backgroundColor = tag.color;
+      tagItem.addEventListener("click", () => filterNotesByTag(tagItem));
+      filterDropdownList.appendChild(tagItem);
+    });
+  }
+
   //render the notes on page load
   renderNotes();
-});
+
+  /**
+   * Filter the notes that are shown by a specific tag. Use the tag item that
+   * is passed in and extract the relevant data from it, and then filter
+   * using that data.
+   * @param {HTMLLIElement} selectedTag
+   */
+  function filterNotesByTag(selectedTag) {
+    const notes = notesAPI.readNotes();
+    const filteredNotes = notes.filter((note) =>
+      note.tags.some(
+        (tag) =>
+          tag.content === selectedTag.textContent &&
+          tag.color === selectedTag.style.backgroundColor
+      )
+    );
+
+    if (filteredNotes.length === 0) {
+      notesContainer.innerHTML = `<h2>Your Journals:</h2>
+      <h3> There are no notes with this tag.</h3>`;
+    } else {
+      renderNotes(filteredNotes);
+    }
+
+    hideFilterDropdown();
+  }
+
+  /**
+   * First checks the user's last saved theme setting, if there is no such setting then defaults to the os/browser theme preference
+   */
+  function loadInitThemePreference() {
+    // check local storage for saved theme
+    const savedTheme = localStorage.getItem("theme");
+
+    // toggle based on saved theme if exists, else toggle based on existing system preferences
+    if (savedTheme && savedTheme == "dark") {
+      modeToggle.checked = true;
+    } else if (savedTheme && savedTheme == "light") {
+      modeToggle.checked = false;
+    } else {
+      modeToggle.checked = themePref.matches;
+    }
+    toggleTheme();
+  }
+
+  /**
+   * Toggles the theme based on the checked status of the theme switch, if checked then switches to dark theme, otherwise switches to light theme
+   */
+  function toggleTheme() {
+    if (modeToggle.checked) {
+      document.body.classList.add("dark-theme");
+      document.body.classList.remove("light-theme");
+      localStorage.setItem("theme", "dark");
+    } else {
+      document.body.classList.add("light-theme");
+      document.body.classList.remove("dark-theme");
+      localStorage.setItem("theme", "light");
+    }
+  }
+
+  return {
+    showNoteEditor,
+    hideNoteEditor,
+    clearNoteEditor,
+    saveActiveNote,
+    deleteActiveNote,
+    deleteNote,
+    renderNotes,
+    filterNotes,
+    createNoteElement,
+    getClosestAncestorEl,
+  };
+}
+
+module.exports = initializeNoteApp;
```

Comment:
what is calling render.js as a module? I don't think it should have any exports as its being treated as a normal scripts file though I may be wrong.

Reply:
That's needed for unit tests.